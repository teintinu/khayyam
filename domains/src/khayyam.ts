import { Bundler } from './bundler'
import { JobManager } from './job'
import { System } from './sys'
import { Workspace } from './workspace'

export interface KhayyamDev {
  readonly workspace: Workspace;
  readonly bundlers: Bundler[]
  readonly jobManager: JobManager
  stop(): void
}

export async function khayyamCI (
  sys: System,
  folder: string,
  jobManager:JobManager
) {
  const workspace = sys.loadWorkspace(folder)
  const build = workspace.walk('all', true, (pkg, bundler) =>
    bundler.build(pkg, jobManager, 'production')
  )
  const test = workspace.walk('all', true, (pkg, bundler) =>
    bundler.test(pkg, jobManager)
  )
  const publish = workspace.walk('all', false, (pkg, bundler) =>
    bundler.publish(pkg, jobManager)
  )
  const lint = workspace.walk('all', false, (pkg, bundler) =>
    bundler.lint(pkg, jobManager)
  )
  const measure = workspace.walk('all', false, (pkg, bundler) =>
    bundler.measure(pkg, jobManager)
  )
  test.depends('each', [build])
  publish.depends('each', [build, test, lint, measure])
  await jobManager.execute()
}

// export function khayyamDev ({
//   jobManager, workspace, watchGlobs
// }:{
//   jobManager:JobManager,
//   path: string
//   bundlers: Bundler[]
//   watchGlobs(globs: string[]): ()=>void
// }) {
//   let watching: Array<()=>void> = []
//   return {
//     stop
//   }
//   function stop () {
//     const close = watching
//     watching = []
//     close.forEach((w) => w())
//   }
//   workspace.bundlers.forEach(wsBundler => {
//     workspace.walk((pkg, bundler) => {
//       const paths = bundler.watch(pkg)
//       const w = watch(paths)
//       watchers.push(w)
//     })
//   })
// //   const queue = queuePromises({
// //     concurrency: cpus().length
// //   })
// //   const km = bootKhayyam(bundlers, queue)
// //   return {
// //     km,
// //     updateWorkspace (workspace: Workspace) {
// //       killWatchers()
// //       km.workspace.packages.forEach((pkg) => {
// //         pkg.bundlers.forEach((bundlerName) => {
// //           const bundler = findBundler(bundlerName)
// //           if (bundler) {
// //             watchers.push(bundler.watch(pkg, (fn) => fn(), queue.promise))
// //           } else {
// //             pkg.progress[bundlerName].update({
// //               state: 'failed',
// //               message: 'Invalid bundle'
// //             })
// //           }
// //         })
// //       })
// //     }
// //   }
// }

// // export function bootKhayyam (
// //   bundlers: Bundler[],
// //   jobManager: JobManager
// // ) : Khayyam {
// //   let workspace: Workspace = {
// //     packages: [],
// //     findBundler,
// //     findPackage
// //   }
// //   let watchers: Job[] = []
// //   const km = {
// //     get workspace () {
// //       return workspace
// //     },
// //     bundlers,
// //     jobs: [],
// //     updateWorkspace (ws: Workspace): void {
// //       workspace = ws
// //     },
// //     updateBundlers (b: Bundler[]): void {
// //       bundlers = b
// //     },
// //     stop () {
// //       killWatchers()
// //     }
// //   }
// //   return km

// //   function killWatchers () {
// //     const old = watchers
// //     watchers = []
// //     old.forEach((w) => w.kill())
// //   }

// // }
