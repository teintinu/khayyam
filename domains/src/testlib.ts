import { createJobManager, JobCallbackArgs, JobManager, Job, Node } from './job'
import { asap, defer } from 'pjobs'
import { Bundler } from './bundler'
import { ByPackage, createWorkspace, Package, WalkedJobs } from './workspace'
import { createProgress } from './progress'
import { System } from './sys'

export const cwd = process.cwd()
const fakeProcess = './scripts/fake-process.js'
const shell = true

export type TreeOfNode = [
  job: string,
  dependencies: TreeOfNode[],
]

export type Logger=ReturnType<typeof createFakeLog>

export function createFakeLog (verbose = false) {
  const logged: string[] = []
  const aOk = defer<void>()
  const bOk = defer<void>()
  const cOk = defer<void>()
  const handleOutput = jest.fn<void, JobCallbackArgs>()
    .mockImplementation(async (job, chunk, error) => {
      const lines = chunk.trim().split('\n')
      lines.forEach((s) => {
        if (s === 'a' || s === 'INPUT: a') setTimeout(aOk.resolve, 100)
        if (s === 'b' || s === 'INPUT: b') setTimeout(bOk.resolve, 100)
        if (s === 'c' || s === 'INPUT: c') setTimeout(cOk.resolve, 100)
        job.progress.update({
          state: 'working',
          message: (error ? ' (error): ' : '') + s
        })
      })
    })
  const jobManager = createJobManager()
  const fsListeners = new Set<()=>void>()
  let tmChanged: any
  const fakeSys: System & { simulateChange():void} = {
    workspaceChanged: fsChanged,
    listenForWorkspaceChanges (callback) {
      fsListeners.add(callback)
      return () => {
        fsListeners.delete(callback)
      }
    },
    getRepository: fsGetRepository,
    simulateChange: fsSimulateChange,
    loadWorkspace,
    notify: fsNotify
  }
  const fakeLogger = {
    jobManager,
    handleOutput,
    aOk,
    bOk,
    cOk,
    fakeSys,
    verbose,
    log (...args: string[]) {
      logged.push(args.join(' '))
      if (verbose) { console.info(args.join(' ')) }
    },
    logged,
    tree () {
      const done: {[id: number]:boolean} = {}
      const tree = jobManager.getTree()
      return tree.map(treeOfNode)
      function treeOfNode (node: Node): TreeOfNode {
        const id = node.job.id
        if (done[id]) {
          return [
            node.job.title + '(rec)',
            []
          ]
        }
        done[id] = true
        const n: TreeOfNode = [
          node.job.title + '=>[' +
          node.dependents.map(d => d.job.title).join() + ']',
          node.dependencies.map(treeOfNode)
        ]
        done[id] = false
        return n
      }
    }
  }
  return fakeLogger
  function fsChanged (): void {
    clearTimeout(tmChanged)
    tmChanged = setTimeout(() => {
      fsListeners.forEach((fn) => asap(fn))
    }, 1000)
  }
  function fsGetRepository (folder: string): string|undefined {
    return 'repo:' + folder
  }
  function fsSimulateChange ():void {
    fsChanged()
  }
  function loadWorkspace (folder:string) {
    const [sBundler, sPkg, sDeps, sLayer, sInvalid] = folder.split('/')
    let bundler:Bundler
    if (sBundler === 'npm') bundler = createFakeBundlerNPM(fakeLogger)
    else throw new Error('fake-fs-invalid-bundle')
    return createFakeWorkspace({
      logger: fakeLogger,
      bundler,
      packages: sPkg as FakePackages,
      deps: sDeps === 'deps',
      layers: sLayer === 'layers',
      invalid: sInvalid === 'invalid'
    })
  }
  function fsNotify (msg: string, pkgName: string) {
    fakeLogger.log('fsNotify:' + msg + ' on package: ' + pkgName)
  }
}

export function createFakeProgress (logger: Logger, title: string, manual: boolean) {
  const pg = createProgress(manual)
  pg.on((status) => {
    logger.log(title + ' ' + status.state + ' ' + status.message)
  })
  return pg
}

export function createFakeJob (logger: Logger, delay: number, title: string, args: string[], manual: boolean = false) {
  logger.log('creating job [' + title + ']')
  const job = logger.jobManager.createJob({
    title,
    cwd,
    command: 'node',
    args: [fakeProcess, String(delay), ...args],
    shell,
    queue: 'sequential',
    manual
  })
  job.progress.on((status) => {
    logger.log('pg ' + title + ': ' + status.state + ' ' + status.message)
  })
  job.listen(logger.handleOutput)
  return job
}

export function create2FakeJob (logger: Logger, delay: number, mainTitle: string, mainArgs: string[], depTitle: string, depArgs: string[]) {
  logger.log('creating main job [' + mainTitle + ']')
  const main = logger.jobManager.createJob({
    title: mainTitle,
    cwd,
    command: 'node',
    args: [fakeProcess, String(delay), ...mainArgs],
    shell,
    queue: 'persistent',
    manual: false
  })
  logger.log('creating dep job [' + depTitle + ']')
  const dep = logger.jobManager.createJob({
    title: depTitle,
    cwd,
    command: 'node',
    args: [fakeProcess, String(delay), ...depArgs],
    shell,
    queue: 'persistent',
    manual: false
  })
  main.depends(dep)
  main.progress.on((status) => {
    logger.log('pg-main ' + status.state + ' ' + status.message)
  })
  dep.progress.on((status) => {
    logger.log('pg-dep ' + status.state + ' ' + status.message)
  })
  main.listen(logger.handleOutput)
  dep.listen(logger.handleOutput)
  return {
    main, dep
  }
}

function createFakeBundlerNPM (logger:Logger) {
  const fakeBundler: Bundler = {
    name: 'fakeBundlerNPM',
    watch (pkg) {
      return ['fakeBundlerNPM']
    },
    build (pkg, jobManager, goal) {
      return createFakeBundlerCommand(pkg, jobManager, 'npm-build-' + goal)
    },
    test (pkg, jobManager) {
      return createFakeBundlerCommand(pkg, jobManager, 'npm-test')
    },
    serve (pkg, jobManager) {
      return createFakeBundlerCommand(pkg, jobManager, 'npm-serve')
    },
    lint (pkg, jobManager) {
      return createFakeBundlerCommand(pkg, jobManager, 'npm-lint')
    },
    publish (pkg, jobManager) {
      return createFakeBundlerCommand(pkg, jobManager, 'npm-publish')
    }
  }
  return fakeBundler
  function createFakeBundlerCommand (pkg: Package, cmdJobManager: JobManager, cmd: string) {
    expect(cmdJobManager).toBe(logger.jobManager)
    return createFakeJob(
      logger,
      100,
      pkg.name + '(' + cmd + ')',
      [
        cmd
      ]
    )
  }
}

type FakePackages='x'|'xy'|'abcde'

function createFakeWorkspace (
  { logger, bundler, packages, deps, layers, invalid }:
  { logger: Logger, bundler: Bundler, packages:FakePackages, deps: boolean; layers: boolean; invalid: boolean }) {
  const pkgs:{
      [n:string]:Package
    } = {}
  for (const n of packages) {
    const pkg: Readonly<Package > = {
      name: n,
      layer: layers ? n : '',
      dependencies: [],
      bundlers: [bundler.name],
      folder: n
    }
    pkgs[n] = pkg
  }
  if (deps) {
    if (packages === 'xy') {
      pkgs.x = {
        ...pkgs.x,
        dependencies: invalid ? ['z'] : ['y']
      }
    }
    if (packages === 'abcde') {
      pkgs.a = {
        ...pkgs.a,
        dependencies: invalid ? ['b', 'c', 'z'] : ['b', 'c']
      }
      pkgs.b = {
        ...pkgs.b,
        dependencies: ['c', 'd']
      }
      pkgs.c = {
        ...pkgs.c,
        dependencies: ['d']
      }
      pkgs.d = {
        ...pkgs.d,
        dependencies: ['e']
      }
    }
  }
  const ws = createWorkspace({
    sys: logger.fakeSys,
    layers: [],
    packages: Object.keys(pkgs).map((n) => pkgs[n]),
    bundlers: [bundler]
  })
  return ws
}

export function jobDebug (job: Job) {
  return job.title + '(' +
    [...job.dependencies]
      .sort((a, b) => a.title.localeCompare(b.title))
      .map(dep => dep.title) +
    ')'
}

export function walkedDebug (walkJobs: WalkedJobs) {
  const ret: ByPackage<unknown> = {}
  const pkgs = Object.keys(walkJobs.jobs).sort()
  pkgs.forEach((pkg) => {
    ret[pkg] = ([...walkJobs.jobs[pkg]] || [])
      .sort((a, b) => a.title.localeCompare(b.title))
      .map(jobDebug)
  })
  return ret
}
