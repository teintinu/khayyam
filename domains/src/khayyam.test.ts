import { khayyamCD, khayyamDev } from './khayyam'
import { createFakeLog } from './testlib'

describe('Khayyam domain', () => {
  describe('CD', () => {
    it('x package', async () => {
      const logger = createFakeLog()
      const p = khayyamCD(
        logger.fakeSys,
        'npm/x/deps',
        logger.jobManager
      )
      expect(logger.logged).toMatchSnapshot('log')
      expect(logger.tree()).toMatchSnapshot('tree')
      await p
    })
    it('xy package', async () => {
      const logger = createFakeLog()
      const p = khayyamCD(
        logger.fakeSys,
        'npm/xy/deps',
        logger.jobManager
      )
      expect(logger.logged).toMatchSnapshot('log')
      expect(logger.tree()).toMatchSnapshot('tree')
      await p
    })
    it('a,b,c,d,e packages', async () => {
      const logger = createFakeLog()
      const p = khayyamCD(
        logger.fakeSys,
        'npm/abcde/deps',
        logger.jobManager
      )
      expect(logger.logged).toMatchSnapshot('log')
      expect(logger.tree()).toMatchSnapshot('tree')
      await p
    }, 10000)
  })
  describe('dev', () => {
    it('x package', async () => {
      const logger = createFakeLog()
      const p = await khayyamDev(
        logger.fakeSys,
        'npm/x/deps',
        logger.jobManager
      )
      expect(logger.logged).toMatchSnapshot('01-log-before build')
      expect(logger.tree()).toMatchSnapshot('01-tree-before build')
      await p.waitBuild()
      expect(logger.logged).toMatchSnapshot('02-log-after build')
      expect(logger.tree()).toMatchSnapshot('02-tree-after build')

      logger.fakeSys.simulateChange('fakeBundlerNPM/x/deps/khayyam.yaml')
      await p.waitBuild()
      expect(logger.logged).toMatchSnapshot('03-log-after rebuild')
      expect(logger.tree()).toMatchSnapshot('03-tree-after rebuild')

      await p.stop()

      expect(logger.logged).toMatchSnapshot('04-log-after stop')
      expect(logger.tree()).toMatchSnapshot('04-tree-after stop')
      expect(logger.sysHandlersState()).toBe('clean')
    }, 15000)
    it('a,b,c,d,e packages', async () => {
      const logger = createFakeLog()
      const p = await khayyamDev(
        logger.fakeSys,
        'npm/abcde/deps',
        logger.jobManager
      )
      expect(logger.logged).toMatchSnapshot('01-log-before build')
      expect(logger.tree()).toMatchSnapshot('01-tree-before build')
      await p.waitBuild()
      expect(logger.logged).toMatchSnapshot('02-log-after build')
      expect(logger.tree()).toMatchSnapshot('02-tree-after build')

      logger.fakeSys.simulateChange('fakeBundlerNPM/abcde/deps/khayyam.yaml')
      await p.waitBuild()
      expect(logger.logged).toMatchSnapshot('03-log-after rebuild')
      expect(logger.tree()).toMatchSnapshot('03-tree-after rebuild')

      await p.stop()

      expect(logger.logged).toMatchSnapshot('04-log-after stop')
      expect(logger.tree()).toMatchSnapshot('04-tree-after stop')
      expect(logger.sysHandlersState()).toBe('clean')
    }, 35000)
  })
})
