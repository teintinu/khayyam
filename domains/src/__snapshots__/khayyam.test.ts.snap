// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Khayyam domain CI a,b,c,d,e packages: log 1`] = `
Array [
  "creating job [e(npm-build-production)]",
  "creating job [d(npm-build-production)]",
  "creating job [c(npm-build-production)]",
  "creating job [b(npm-build-production)]",
  "creating job [a(npm-build-production)]",
  "creating job [e(npm-test)]",
  "creating job [d(npm-test)]",
  "creating job [c(npm-test)]",
  "creating job [b(npm-test)]",
  "creating job [a(npm-test)]",
  "creating job [e(npm-publish)]",
  "creating job [d(npm-publish)]",
  "creating job [c(npm-publish)]",
  "creating job [b(npm-publish)]",
  "creating job [a(npm-publish)]",
  "creating job [e(npm-lint)]",
  "creating job [d(npm-lint)]",
  "creating job [c(npm-lint)]",
  "creating job [b(npm-lint)]",
  "creating job [a(npm-lint)]",
  "creating job [e(npm-measure)]",
  "creating job [d(npm-measure)]",
  "creating job [c(npm-measure)]",
  "creating job [b(npm-measure)]",
  "creating job [a(npm-measure)]",
]
`;

exports[`Khayyam domain CI a,b,c,d,e packages: tree 1`] = `
"e(npm-publish)
  dependents:
  dependencies:
    e(npm-build-production)
      dependents:
        d(npm-build-production)
        e(npm-test)
        e(npm-publish)
      dependencies:
    e(npm-test)
      dependents:
        d(npm-test)
        e(npm-publish)
      dependencies:
        e(npm-build-production)
          dependents:
            d(npm-build-production)
            e(npm-test)
            e(npm-publish)
          dependencies:
    e(npm-lint)
      dependents:
        e(npm-publish)
      dependencies:
    e(npm-measure)
      dependents:
        e(npm-publish)
      dependencies:
d(npm-publish)
  dependents:
  dependencies:
    d(npm-build-production)
      dependents:
        c(npm-build-production)
        b(npm-build-production)
        d(npm-test)
        d(npm-publish)
      dependencies:
        e(npm-build-production)
          dependents:
            d(npm-build-production)
            e(npm-test)
            e(npm-publish)
          dependencies:
    d(npm-test)
      dependents:
        c(npm-test)
        b(npm-test)
        d(npm-publish)
      dependencies:
        e(npm-test)
          dependents:
            d(npm-test)
            e(npm-publish)
          dependencies:
            e(npm-build-production)
              dependents:
                d(npm-build-production)
                e(npm-test)
                e(npm-publish)
              dependencies:
        d(npm-build-production)
          dependents:
            c(npm-build-production)
            b(npm-build-production)
            d(npm-test)
            d(npm-publish)
          dependencies:
            e(npm-build-production)
              dependents:
                d(npm-build-production)
                e(npm-test)
                e(npm-publish)
              dependencies:
    d(npm-lint)
      dependents:
        d(npm-publish)
      dependencies:
    d(npm-measure)
      dependents:
        d(npm-publish)
      dependencies:
c(npm-publish)
  dependents:
  dependencies:
    c(npm-build-production)
      dependents:
        b(npm-build-production)
        a(npm-build-production)
        c(npm-test)
        c(npm-publish)
      dependencies:
        d(npm-build-production)
          dependents:
            c(npm-build-production)
            b(npm-build-production)
            d(npm-test)
            d(npm-publish)
          dependencies:
            e(npm-build-production)
              dependents:
                d(npm-build-production)
                e(npm-test)
                e(npm-publish)
              dependencies:
    c(npm-test)
      dependents:
        b(npm-test)
        a(npm-test)
        c(npm-publish)
      dependencies:
        d(npm-test)
          dependents:
            c(npm-test)
            b(npm-test)
            d(npm-publish)
          dependencies:
            e(npm-test)
              dependents:
                d(npm-test)
                e(npm-publish)
              dependencies:
                e(npm-build-production)
                  dependents:
                    d(npm-build-production)
                    e(npm-test)
                    e(npm-publish)
                  dependencies:
            d(npm-build-production)
              dependents:
                c(npm-build-production)
                b(npm-build-production)
                d(npm-test)
                d(npm-publish)
              dependencies:
                e(npm-build-production)
                  dependents:
                    d(npm-build-production)
                    e(npm-test)
                    e(npm-publish)
                  dependencies:
        c(npm-build-production)
          dependents:
            b(npm-build-production)
            a(npm-build-production)
            c(npm-test)
            c(npm-publish)
          dependencies:
            d(npm-build-production)
              dependents:
                c(npm-build-production)
                b(npm-build-production)
                d(npm-test)
                d(npm-publish)
              dependencies:
                e(npm-build-production)
                  dependents:
                    d(npm-build-production)
                    e(npm-test)
                    e(npm-publish)
                  dependencies:
    c(npm-lint)
      dependents:
        c(npm-publish)
      dependencies:
    c(npm-measure)
      dependents:
        c(npm-publish)
      dependencies:
b(npm-publish)
  dependents:
  dependencies:
    b(npm-build-production)
      dependents:
        a(npm-build-production)
        b(npm-test)
        b(npm-publish)
      dependencies:
        c(npm-build-production)
          dependents:
            b(npm-build-production)
            a(npm-build-production)
            c(npm-test)
            c(npm-publish)
          dependencies:
            d(npm-build-production)
              dependents:
                c(npm-build-production)
                b(npm-build-production)
                d(npm-test)
                d(npm-publish)
              dependencies:
                e(npm-build-production)
                  dependents:
                    d(npm-build-production)
                    e(npm-test)
                    e(npm-publish)
                  dependencies:
        d(npm-build-production)
          dependents:
            c(npm-build-production)
            b(npm-build-production)
            d(npm-test)
            d(npm-publish)
          dependencies:
            e(npm-build-production)
              dependents:
                d(npm-build-production)
                e(npm-test)
                e(npm-publish)
              dependencies:
    b(npm-test)
      dependents:
        a(npm-test)
        b(npm-publish)
      dependencies:
        c(npm-test)
          dependents:
            b(npm-test)
            a(npm-test)
            c(npm-publish)
          dependencies:
            d(npm-test)
              dependents:
                c(npm-test)
                b(npm-test)
                d(npm-publish)
              dependencies:
                e(npm-test)
                  dependents:
                    d(npm-test)
                    e(npm-publish)
                  dependencies:
                    e(npm-build-production)
                      dependents:
                        d(npm-build-production)
                        e(npm-test)
                        e(npm-publish)
                      dependencies:
                d(npm-build-production)
                  dependents:
                    c(npm-build-production)
                    b(npm-build-production)
                    d(npm-test)
                    d(npm-publish)
                  dependencies:
                    e(npm-build-production)
                      dependents:
                        d(npm-build-production)
                        e(npm-test)
                        e(npm-publish)
                      dependencies:
            c(npm-build-production)
              dependents:
                b(npm-build-production)
                a(npm-build-production)
                c(npm-test)
                c(npm-publish)
              dependencies:
                d(npm-build-production)
                  dependents:
                    c(npm-build-production)
                    b(npm-build-production)
                    d(npm-test)
                    d(npm-publish)
                  dependencies:
                    e(npm-build-production)
                      dependents:
                        d(npm-build-production)
                        e(npm-test)
                        e(npm-publish)
                      dependencies:
        d(npm-test)
          dependents:
            c(npm-test)
            b(npm-test)
            d(npm-publish)
          dependencies:
            e(npm-test)
              dependents:
                d(npm-test)
                e(npm-publish)
              dependencies:
                e(npm-build-production)
                  dependents:
                    d(npm-build-production)
                    e(npm-test)
                    e(npm-publish)
                  dependencies:
            d(npm-build-production)
              dependents:
                c(npm-build-production)
                b(npm-build-production)
                d(npm-test)
                d(npm-publish)
              dependencies:
                e(npm-build-production)
                  dependents:
                    d(npm-build-production)
                    e(npm-test)
                    e(npm-publish)
                  dependencies:
        b(npm-build-production)
          dependents:
            a(npm-build-production)
            b(npm-test)
            b(npm-publish)
          dependencies:
            c(npm-build-production)
              dependents:
                b(npm-build-production)
                a(npm-build-production)
                c(npm-test)
                c(npm-publish)
              dependencies:
                d(npm-build-production)
                  dependents:
                    c(npm-build-production)
                    b(npm-build-production)
                    d(npm-test)
                    d(npm-publish)
                  dependencies:
                    e(npm-build-production)
                      dependents:
                        d(npm-build-production)
                        e(npm-test)
                        e(npm-publish)
                      dependencies:
            d(npm-build-production)
              dependents:
                c(npm-build-production)
                b(npm-build-production)
                d(npm-test)
                d(npm-publish)
              dependencies:
                e(npm-build-production)
                  dependents:
                    d(npm-build-production)
                    e(npm-test)
                    e(npm-publish)
                  dependencies:
    b(npm-lint)
      dependents:
        b(npm-publish)
      dependencies:
    b(npm-measure)
      dependents:
        b(npm-publish)
      dependencies:
a(npm-publish)
  dependents:
  dependencies:
    a(npm-build-production)
      dependents:
        a(npm-test)
        a(npm-publish)
      dependencies:
        b(npm-build-production)
          dependents:
            a(npm-build-production)
            b(npm-test)
            b(npm-publish)
          dependencies:
            c(npm-build-production)
              dependents:
                b(npm-build-production)
                a(npm-build-production)
                c(npm-test)
                c(npm-publish)
              dependencies:
                d(npm-build-production)
                  dependents:
                    c(npm-build-production)
                    b(npm-build-production)
                    d(npm-test)
                    d(npm-publish)
                  dependencies:
                    e(npm-build-production)
                      dependents:
                        d(npm-build-production)
                        e(npm-test)
                        e(npm-publish)
                      dependencies:
            d(npm-build-production)
              dependents:
                c(npm-build-production)
                b(npm-build-production)
                d(npm-test)
                d(npm-publish)
              dependencies:
                e(npm-build-production)
                  dependents:
                    d(npm-build-production)
                    e(npm-test)
                    e(npm-publish)
                  dependencies:
        c(npm-build-production)
          dependents:
            b(npm-build-production)
            a(npm-build-production)
            c(npm-test)
            c(npm-publish)
          dependencies:
            d(npm-build-production)
              dependents:
                c(npm-build-production)
                b(npm-build-production)
                d(npm-test)
                d(npm-publish)
              dependencies:
                e(npm-build-production)
                  dependents:
                    d(npm-build-production)
                    e(npm-test)
                    e(npm-publish)
                  dependencies:
    a(npm-test)
      dependents:
        a(npm-publish)
      dependencies:
        b(npm-test)
          dependents:
            a(npm-test)
            b(npm-publish)
          dependencies:
            c(npm-test)
              dependents:
                b(npm-test)
                a(npm-test)
                c(npm-publish)
              dependencies:
                d(npm-test)
                  dependents:
                    c(npm-test)
                    b(npm-test)
                    d(npm-publish)
                  dependencies:
                    e(npm-test)
                      dependents:
                        d(npm-test)
                        e(npm-publish)
                      dependencies:
                        e(npm-build-production)
                          dependents:
                            d(npm-build-production)
                            e(npm-test)
                            e(npm-publish)
                          dependencies:
                    d(npm-build-production)
                      dependents:
                        c(npm-build-production)
                        b(npm-build-production)
                        d(npm-test)
                        d(npm-publish)
                      dependencies:
                        e(npm-build-production)
                          dependents:
                            d(npm-build-production)
                            e(npm-test)
                            e(npm-publish)
                          dependencies:
                c(npm-build-production)
                  dependents:
                    b(npm-build-production)
                    a(npm-build-production)
                    c(npm-test)
                    c(npm-publish)
                  dependencies:
                    d(npm-build-production)
                      dependents:
                        c(npm-build-production)
                        b(npm-build-production)
                        d(npm-test)
                        d(npm-publish)
                      dependencies:
                        e(npm-build-production)
                          dependents:
                            d(npm-build-production)
                            e(npm-test)
                            e(npm-publish)
                          dependencies:
            d(npm-test)
              dependents:
                c(npm-test)
                b(npm-test)
                d(npm-publish)
              dependencies:
                e(npm-test)
                  dependents:
                    d(npm-test)
                    e(npm-publish)
                  dependencies:
                    e(npm-build-production)
                      dependents:
                        d(npm-build-production)
                        e(npm-test)
                        e(npm-publish)
                      dependencies:
                d(npm-build-production)
                  dependents:
                    c(npm-build-production)
                    b(npm-build-production)
                    d(npm-test)
                    d(npm-publish)
                  dependencies:
                    e(npm-build-production)
                      dependents:
                        d(npm-build-production)
                        e(npm-test)
                        e(npm-publish)
                      dependencies:
            b(npm-build-production)
              dependents:
                a(npm-build-production)
                b(npm-test)
                b(npm-publish)
              dependencies:
                c(npm-build-production)
                  dependents:
                    b(npm-build-production)
                    a(npm-build-production)
                    c(npm-test)
                    c(npm-publish)
                  dependencies:
                    d(npm-build-production)
                      dependents:
                        c(npm-build-production)
                        b(npm-build-production)
                        d(npm-test)
                        d(npm-publish)
                      dependencies:
                        e(npm-build-production)
                          dependents:
                            d(npm-build-production)
                            e(npm-test)
                            e(npm-publish)
                          dependencies:
                d(npm-build-production)
                  dependents:
                    c(npm-build-production)
                    b(npm-build-production)
                    d(npm-test)
                    d(npm-publish)
                  dependencies:
                    e(npm-build-production)
                      dependents:
                        d(npm-build-production)
                        e(npm-test)
                        e(npm-publish)
                      dependencies:
        c(npm-test)
          dependents:
            b(npm-test)
            a(npm-test)
            c(npm-publish)
          dependencies:
            d(npm-test)
              dependents:
                c(npm-test)
                b(npm-test)
                d(npm-publish)
              dependencies:
                e(npm-test)
                  dependents:
                    d(npm-test)
                    e(npm-publish)
                  dependencies:
                    e(npm-build-production)
                      dependents:
                        d(npm-build-production)
                        e(npm-test)
                        e(npm-publish)
                      dependencies:
                d(npm-build-production)
                  dependents:
                    c(npm-build-production)
                    b(npm-build-production)
                    d(npm-test)
                    d(npm-publish)
                  dependencies:
                    e(npm-build-production)
                      dependents:
                        d(npm-build-production)
                        e(npm-test)
                        e(npm-publish)
                      dependencies:
            c(npm-build-production)
              dependents:
                b(npm-build-production)
                a(npm-build-production)
                c(npm-test)
                c(npm-publish)
              dependencies:
                d(npm-build-production)
                  dependents:
                    c(npm-build-production)
                    b(npm-build-production)
                    d(npm-test)
                    d(npm-publish)
                  dependencies:
                    e(npm-build-production)
                      dependents:
                        d(npm-build-production)
                        e(npm-test)
                        e(npm-publish)
                      dependencies:
        a(npm-build-production)
          dependents:
            a(npm-test)
            a(npm-publish)
          dependencies:
            b(npm-build-production)
              dependents:
                a(npm-build-production)
                b(npm-test)
                b(npm-publish)
              dependencies:
                c(npm-build-production)
                  dependents:
                    b(npm-build-production)
                    a(npm-build-production)
                    c(npm-test)
                    c(npm-publish)
                  dependencies:
                    d(npm-build-production)
                      dependents:
                        c(npm-build-production)
                        b(npm-build-production)
                        d(npm-test)
                        d(npm-publish)
                      dependencies:
                        e(npm-build-production)
                          dependents:
                            d(npm-build-production)
                            e(npm-test)
                            e(npm-publish)
                          dependencies:
                d(npm-build-production)
                  dependents:
                    c(npm-build-production)
                    b(npm-build-production)
                    d(npm-test)
                    d(npm-publish)
                  dependencies:
                    e(npm-build-production)
                      dependents:
                        d(npm-build-production)
                        e(npm-test)
                        e(npm-publish)
                      dependencies:
            c(npm-build-production)
              dependents:
                b(npm-build-production)
                a(npm-build-production)
                c(npm-test)
                c(npm-publish)
              dependencies:
                d(npm-build-production)
                  dependents:
                    c(npm-build-production)
                    b(npm-build-production)
                    d(npm-test)
                    d(npm-publish)
                  dependencies:
                    e(npm-build-production)
                      dependents:
                        d(npm-build-production)
                        e(npm-test)
                        e(npm-publish)
                      dependencies:
    a(npm-lint)
      dependents:
        a(npm-publish)
      dependencies:
    a(npm-measure)
      dependents:
        a(npm-publish)
      dependencies:"
`;

exports[`Khayyam domain CI x package: log 1`] = `
Array [
  "creating job [x(npm-build-production)]",
  "creating job [x(npm-test)]",
  "creating job [x(npm-publish)]",
  "creating job [x(npm-lint)]",
  "creating job [x(npm-measure)]",
]
`;

exports[`Khayyam domain CI x package: log 2`] = `
Array [
  "creating job [y(npm-build-production)]",
  "creating job [x(npm-build-production)]",
  "creating job [y(npm-test)]",
  "creating job [x(npm-test)]",
  "creating job [y(npm-publish)]",
  "creating job [x(npm-publish)]",
  "creating job [y(npm-lint)]",
  "creating job [x(npm-lint)]",
  "creating job [y(npm-measure)]",
  "creating job [x(npm-measure)]",
]
`;

exports[`Khayyam domain CI x package: tree 1`] = `
"x(npm-publish)
  dependents:
  dependencies:
    x(npm-build-production)
      dependents:
        x(npm-test)
        x(npm-publish)
      dependencies:
    x(npm-test)
      dependents:
        x(npm-publish)
      dependencies:
        x(npm-build-production)
          dependents:
            x(npm-test)
            x(npm-publish)
          dependencies:
    x(npm-lint)
      dependents:
        x(npm-publish)
      dependencies:
    x(npm-measure)
      dependents:
        x(npm-publish)
      dependencies:"
`;

exports[`Khayyam domain CI x package: tree 2`] = `
"y(npm-publish)
  dependents:
  dependencies:
    y(npm-build-production)
      dependents:
        x(npm-build-production)
        y(npm-test)
        y(npm-publish)
      dependencies:
    y(npm-test)
      dependents:
        x(npm-test)
        y(npm-publish)
      dependencies:
        y(npm-build-production)
          dependents:
            x(npm-build-production)
            y(npm-test)
            y(npm-publish)
          dependencies:
    y(npm-lint)
      dependents:
        y(npm-publish)
      dependencies:
    y(npm-measure)
      dependents:
        y(npm-publish)
      dependencies:
x(npm-publish)
  dependents:
  dependencies:
    x(npm-build-production)
      dependents:
        x(npm-test)
        x(npm-publish)
      dependencies:
        y(npm-build-production)
          dependents:
            x(npm-build-production)
            y(npm-test)
            y(npm-publish)
          dependencies:
    x(npm-test)
      dependents:
        x(npm-publish)
      dependencies:
        y(npm-test)
          dependents:
            x(npm-test)
            y(npm-publish)
          dependencies:
            y(npm-build-production)
              dependents:
                x(npm-build-production)
                y(npm-test)
                y(npm-publish)
              dependencies:
        x(npm-build-production)
          dependents:
            x(npm-test)
            x(npm-publish)
          dependencies:
            y(npm-build-production)
              dependents:
                x(npm-build-production)
                y(npm-test)
                y(npm-publish)
              dependencies:
    x(npm-lint)
      dependents:
        x(npm-publish)
      dependencies:
    x(npm-measure)
      dependents:
        x(npm-publish)
      dependencies:"
`;
